ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -1610,41 +1615,42 @@ withConnectionManager ConnectionManagerArguments {
                             writeTVar connVar connState'
                             return
                               ( state
-                              , Right [ mkTransition connState connState'
-                                      , Transition (Known connState') Unknown
-                                      ]
+                              , [ mkTransition connState connState'
+                                , Transition (Known connState') Unknown
+                                ]
                               )
 
                           -- Current connVar.
-                          Just mutableConnState' ->
-                            -- If accept call returned first than connect then
-                            -- the connVar will be replaced. If it was
-                            -- replaced then we do not need to do anything.
-                            -- Otherwise, we need to remove the connVar from
-                            -- the state and trace accordingly.
-                            if mutableConnState' == mutableConnState
-                               then do
-                                connState <- readTVar connVar
+                          Just mutableConnState' -> do
+                            connState <- readTVar connVar
+                            case connState of
+                              -- Update the state only if the connection was in
+                              -- 'ReservedOutboundState'.  This covers the case
+                              -- when we connect to our selves, in which
+                              -- case: we first set the connection state to
+                              -- `ReservedOutboundState`, then race connect
+                              -- & accept calls.  If the connection was
+                              -- accepted it, it will use the same
+                              -- 'MutableConnState', and if the inbound side is
+                              -- using the connection the state will be
+                              -- different than `ReservedOutboundState`.
+                              ReservedOutboundState | mutableConnState' == mutableConnState -> do
```

Comment:
It seems that the only other possible state here is `UnnegotiatedSt Inbound`; handshake will not happen because we start the connection handler only after `connect` returns.